File: 14.Longest Common Prefix/Sidney/LongestCommonPrefix.c

int allhave(char** strs, int strsSize, int index)
{
    char tmp = strs[0][index];
    for(int i = 0; i < strsSize; i++) {
        if(tmp != strs[i][index]) {
            return 0;
        }
    }
    return 1;
}

char* longestCommonPrefix(char** strs, int strsSize) {
    int min = 2147483647;
    int index = 0;
    for(int i = 0; i < strsSize; i++) {
        if(strlen(strs[i]) < min) {
            min = strlen(strs[i]);
        }
    }
    
    if(min == 2147483647 || min <= 0) {
        return "";
    }
    
    char* ans = (char*)malloc(min);
    if(!ans) {
        return NULL;
    }
    memset(ans, 0, min);
    for(int i = 0; i < min; i++) {
        if(allhave(strs, strsSize, i)) {
            ans[index] = strs[0][i];
            index++;
            ans[index] = 0;
        }
        else {
            break;
        }
    }
    

    return ans;
}

File: 014.Longest Common Prefix/MarkCSIE/LongestCommonPrefix/src/LongestCommonPrefix.java

public class LongestCommonPrefix {

    public static void main(String[] args) {
        LongestCommonPrefix test = new LongestCommonPrefix();
        String[] testCase1 = {"123", "12"};
        System.out.println(test.longestCommonPrefix(testCase1));
    }

    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0) {
            return "";
        } else if (strs.length == 1) {
            return strs[0];
        }

        String lcp = "";
        while (true) {
            int i = 0;
            int index = lcp.length();
            for (i = 0; i < strs.length - 1; i++) {
                if (index >= strs[i].length() || index >= strs[i + 1].length() || strs[i].charAt(index) != strs[i + 1].charAt(index)) {
                    return lcp;
                }
            }
            if (i == strs.length - 1) {
                lcp += strs[0].charAt(index);
            }
        }
    }
}
